feat: Support serializing and deserializing LoRA adapters

sangstar opened the pull request:
# Serializing and deserializing LoRA adapters using `tensorizer`
This PR allows LoRA adapters to be serialized and deserialized using `tensorizer`. A test is added to confirm this and serializing LoRA files can be done with the same script that can save vLLM models in `examples/tensorize_vllm_model.py`.

## Summary of changes

- Adjusted the CLI args for `.buildkite/test-pipeline.yaml`'s invocation of the `tensorize_vllm_model.py` example script to additionally save and load a LoRA adapter, testing a model generation after deserialization to confirm the LoRA adapter was loaded properly
- Added `--lora-path` to `tensorize_vllm_model.py` as a base argparser argument that allows a user to specify the HuggingFace reference ID for a LoRA adapter, which can be either serialized or deserialized depending on whether the `serialize` or `deserialize` subparser is indicated.
- Added `test_serialize_and_deserialize_lora` to `test_tensorizer.py` testing saving and loading LoRA adapters with `tensorizer`.
- Allows deserializing a LoRA adapter by allowing `TensorizerConfig` to be passed in as a kwarg to `LoRARequest`. When this is done, the LoRA tensors are assumed to be in `tensorizer` format and deserialized according to the parameters given in the `TensorizerConfig` provided.


**Note**: Noticed a few things that need hotfixes after putting up this PR. Stamped `TODO`s on these and working on them now while waiting for reviews. If you still see any `TODO`'s (which means I haven't addressed them yet) please do feel free to comment on them.

~NB: In addition, **an extra planned change** will be added to their test pipeline, which will have OpenAI API inference engine `tensorizer` support tests to ensure these can be part of fast-check for forward-compatibility. As such, the state of this PR is not final and will warrant a further review once that piece is added (may be more appropriate to draft this but a review is warranted on the current state).~ This has been added in [cc3f984](https://github.com/coreweave/vllm/pull/3/commits/cc3f98474f0cad362c2556ea37484d5f3e36630b)

sangstar committed:
feat: Add initial conceptual logic for loading LoRA via tensorizer

sangstar committed:
fix: Make passing of Tensorizer-LoRA args more logical

sangstar committed:
feat: Add serializing LoRA adapter in `tensorize_vllm_model.py`

sangstar committed:
feat: Prepare example script for build pipeline

sangstar committed:
chore: Run `format.sh`

sangstar committed:
chore: Change test name

sangstar committed:
chore: Rename `lora_model` to `lora` in test name

sangstar committed:
chore: Add `TODO` about fixing json upload machinery

sangstar committed:
feat: Add `Entrypoints` test

sangstar committed:
fix: Use `open_stream` to serialize `adapter_config.json`

sangstar committed:
docs: Justify non-generic filenames for adapter tensors and json

sangstar committed:
tests: Update unit test and add docstring for `tensorize_lora_adapter`

Rexwang8 commented on .buildkite/test-pipeline.yaml:
```diff
@@ -196,7 +196,7 @@ steps:
     - python3 offline_inference_vision_language.py
     - python3 offline_inference_vision_language_multi_image.py
-    - python3 tensorize_vllm_model.py --model facebook/opt-125m serialize --serialized-directory /tmp/ --suffix v1 && python3 tensorize_vllm_model.py --model facebook/opt-125m deserialize --path-to-tensors /tmp/vllm/facebook/opt-125m/v1/model.tensors
+    - python3 tensorize_vllm_model.py --model meta-llama/Llama-2-7b-hf --lora-path yard1/llama-2-7b-sql-lora-test serialize --serialized-directory /tmp/ --suffix v1 && python3 tensorize_vllm_model.py --model meta-llama/Llama-2-7b-hf --lora-path yard1/llama-2-7b-sql-lora-test deserialize --path-to-tensors /tmp/vllm/meta-llama/Llama-2-7b-hf/v1/model.tensors
```
Can you explain the reasoning behind changing the model here?

Rexwang8 commented on vllm/lora/models.py:
```diff
@@ -175,19 +175,20 @@ def from_lora_tensors(
-        embedding_padding_modules: Optional[List[str]] = None,
-    ) -> "LoRAModel":
+    def from_local_checkpoint(cls,
+                              lora_dir: str,
```
It looks like the formatting here is changed from the base?

Rexwang8 reviewed (approved):
Looks good! Tests pass, code looks fine. Good job!!

sangstar committed:
chore: Add note on broken change

sangstar committed:
chore: Add TODO

sangstar committed:
fix: Get LoRA tensorizer test passing again

sangstar committed:
fix: Remove env var dependency, ensure string compliance for tests for `tensorizer_uri`

sangstar committed:
style: Remove redundant line

sangstar committed:
fix: Override usage of `lora_path` to coincide with keeping LoRA artifacts in `tensorizer_dir`

sangstar committed:
fix: Update entrypoint test, correctly parse `lora_tensor_path`, remove >Python3.10 syntax

sangstar committed:
chore: Apply linter fixes from pre-commit hooks

sangstar committed:
chore: Remove `tensorizer_config` type annotation; clarify with comment

Eta0 commented on tests/entrypoints/openai/test_tensorizer_entrypoint.py:
```diff
@@ -0,0 +1,114 @@
+
+    yield tmp_dir
+
+    cleanup()
```
A better pattern for this is usually to use `try` + `finally`, but in this case, `tempfile.TemporaryDirectory` is a context manager, so you can achieve the same effect with a `with` statement:
```suggestion
def tmp_dir():
    with tempfile.TemporaryDirectory() as path:
        yield path
```
(Note: this yields the equivalent of just `tmp_dir.name` as a `str` instead of the full stateful object)

sangstar replied:
I guess `try + finally` is pretty functionally equivalent to a context manager due to having a guaranteed cleanup?

Eta0 commented on tests/entrypoints/openai/test_tensorizer_entrypoint.py:
```diff
@@ -0,0 +1,114 @@
+
+@pytest.fixture(scope="module")
+def tensorize_model_and_lora(tmp_dir):
+    model_uri = tmp_dir.name + "/model.tensors"
```
You repeatedly use `tmp_dir` only to use it in the expression `tmp_dir.name + "/model.tensors"`, so it seems like it would make more sense to have another fixture that returns this string directly.

sangstar committed:
chore: Use `typing.TYPE_CHECKING` to import `TensorizerConfig` for annotation

Eta0 commented on tests/entrypoints/openai/test_tensorizer_entrypoint.py:
```diff
@@ -0,0 +1,114 @@
+def tensorize_model_and_lora(tmp_dir):
+    model_uri = tmp_dir.name + "/model.tensors"
+    tensorizer_config = TensorizerConfig(tensorizer_uri=model_uri)
+    args = EngineArgs(model=MODEL_NAME, )
```
Extraneous comma and space.
```suggestion
    args = EngineArgs(model=MODEL_NAME)
```

sangstar replied:
Thanks for the catch. I wonder why the linters never spotted this?

Eta0 commented on tests/entrypoints/openai/test_tensorizer_entrypoint.py:
```diff
@@ -0,0 +1,114 @@
+                                     max_tokens=256,
+                                     stop=["[/assistant]"])
+
+    prompts = [
+        "[user] Write a SQL query to answer the question based on the table schema.\n\n context: CREATE TABLE table_name_74 (icao VARCHAR, airport VARCHAR)\n\n question: Name the ICAO for lilongwe international airport [/user] [assistant]",  # noqa: E501
+        "[user] Write a SQL query to answer the question based on the table schema.\n\n context: CREATE TABLE table_name_11 (nationality VARCHAR, elector VARCHAR)\n\n question: When Anchero Pantaleone was the elector what is under nationality? [/user] [assistant]",  # noqa: E501
+    ]
```
You should split these strings up instead of disabling the inspection here.

sangstar replied:
[This was directly scraped from the LoRA tests.](https://github.com/vllm-project/vllm/blob/ed6e9075d31e32c8548b480a47d1ffb77da1f54c/tests/lora/test_llama_tp.py#L35-L42)

Happy to adjust this, unless there's any risk of it being perceived as stuffy of me to try to use their test prompts but pretty it up, but surely that's overthinking.

sangstar committed:
chore: Stringify `TensorizerConfig` annotation

Eta0 commented on tests/entrypoints/openai/test_tensorizer_entrypoint.py:
```diff
@@ -0,0 +1,114 @@
+    assert completion.model == MODEL_NAME
+    assert len(completion.choices) == 1
+    assert len(completion.choices[0].text) >= 5
+    assert completion.choices[0].finish_reason == "length"
+    assert completion.usage == openai.types.CompletionUsage(
+        completion_tokens=5, prompt_tokens=6, total_tokens=11)
```
Is it valid to have any other potential finish reasons occur here?

sangstar replied:
There wasn't much thought put in to this, frankly. It's reused the same couple lines of code from `test_single_completion` from `test_completion.py` and `test_encoder_decoder.py`, so I deferred to those scripts to justify the correctness of those particular couple of assertions.

Eta0 commented on vllm/lora/models.py:
```diff
@@ -215,9 +216,22 @@ def from_local_checkpoint(
+                lora_tensor_path,
+                dtype=tensorizer_config.dtype,
+                device=f'cuda:{torch.cuda.current_device()}',
+                **tensorizer_args.deserializer_params)
```
Tensorizer's default is to use `torch.device("cuda")` if CUDA is available (and `torch.device("cpu")` otherwise), which will generally correspond to the same device as `cuda:{torch.cuda.current_device()}` unless you expect the contextual current device to change for some reason in the middle of reading the tensors. In other words, you shouldn't have to specify this here.
```suggestion
            tensors = TensorDeserializer(
                lora_tensor_path,
                dtype=tensorizer_config.dtype,
                **tensorizer_args.deserializer_params)
```

Eta0 commented on vllm/lora/models.py:
```diff
@@ -238,8 +252,8 @@ def from_local_checkpoint(
-                        f" Please verify that the loaded LoRA module is correct"
-                    )
+                        f" Please verify that the loaded LoRA module "
+                        f"is correct")
```
Not sure why this was changed
```suggestion
                        f" Please verify that the loaded LoRA module is correct"
                    )
```

sangstar replied:
Not sure either.

Eta0 commented on vllm/lora/models.py:
```diff
@@ -215,9 +216,22 @@ def from_local_checkpoint(
         new_embeddings_bin_file_path = os.path.join(lora_dir,
                                                     "new_embeddings.bin")
 
+        tensorizer_config = kwargs.get("tensorizer_config")
```
Why is this using `kwargs` instead of an argument with a default value? Adding `kwargs` to a function is significant since it will start silently accepting and ignoring incorrect parameter names passed to the function if you don't also go out of your way to check for extraneous entries in it. Is there a reason you need to use `**kwargs` over `tensorizer_config=None`?

sangstar replied:
The motivation was to along the lines of "I want to include my tensorizer hook, but I don't want to be as intrusive as to ask the developers to include my `tensorizer_config` as directly as an acknowledged parameter in their LoRA API with a docstring and everything. I'll instead only be intrusive enough as to ask them to allow `**kwargs`, setting a precedent for others to hook in their own stuff unintrusively.".

This has made me see though that this is the opposite of unintrusive; it's harming the API itself to allow `**kwargs` like this and making the LoRA machinery harder to maintain and setting a precedent of LoRA logic having to handle any number of random, undocumented parameters. I'll explicitly introduce `tensorizer_config` as a parameter.

Eta0 commented on vllm/lora/peft_helper.py:
```diff
@@ -90,11 +90,25 @@ def from_dict(cls, config_dict: dict) -> "PEFTHelper":
+                       max_position_embeddings: Optional[int],
+                       **kwargs) -> "PEFTHelper":
         lora_config_path = os.path.join(lora_path, "adapter_config.json")
 
-        with open(lora_config_path) as f:
-            config = json.load(f)
+        tensorizer_config = kwargs.get("tensorizer_config")
```
Same comment about adding `kwargs`. Can this be replaced with a normal parameter with a default value?

sangstar replied:
Yes. Will do.

Eta0 commented on vllm/lora/peft_helper.py:
```diff
@@ -90,11 +90,25 @@ def from_dict(cls, config_dict: dict) -> "PEFTHelper":
+        tensorizer_config = kwargs.get("tensorizer_config")
+        if tensorizer_config:
+            tensorizer_args = tensorizer_config._construct_tensorizer_args()
+            from tensorizer.stream_io import open_stream
+            lora_config_path = os.path.join(tensorizer_config.tensorizer_dir,
+                                            "adapter_config.json")
+            with open_stream(lora_config_path,
+                             mode="rb",
+                             **tensorizer_args.stream_params) as f:
```
It seems misleading that a method named `from_local_dir` is explicitly given the opportunity here to load something from the network rather than from a local dir.

sangstar replied:
I agree that it's misleading. It was, however, the most logical place to inject hooks for loading with tensorizer. I think the code is where it should be if I ignore the name of the function, as functionally it's proven to be appropriately placed (as this is the official "config loading" piece, albeit meant to be local). I'm not really sure what to do about this that would make sense. I don't think renaming the class method would be appropriate, and I don't think this code should be anywhere else. Interested to hear your thoughts on this.

Eta0 commented on vllm/lora/worker_manager.py:
```diff
@@ -99,7 +99,10 @@ def _load_adapter(self, lora_request: LoRARequest) -> LoRAModel:
-                lora_path, self.max_position_embeddings)
+                lora_path,
+                self.max_position_embeddings,
+                tensorizer_config=getattr(lora_request, "tensorizer_config",
+                                          None))
```
You defined the `LoRARequest` class to have a default value of `None` for `lora_request.tensorizer_config`. Is there a reason (e.g. metaclass witchcraft) the attribute might not be defined here in spite of that that motivated using `getattr`?

sangstar replied:
I'm a bit hazy on my justification for this. I think initially I didn't give it a default value of `None`, but something about not doing this seemed to cause it to infer its type weirdly or something (some `msgspec.Struct` weirdness or something maybe). I can take a second look, but once it was happy when I made `None` its default value I mostly shrugged and carried on.

Eta0 commented on vllm/model_executor/model_loader/tensorizer.py:
```diff
@@ -11,6 +12,8 @@
 from typing import BinaryIO, Generator, Optional, Tuple, Type, Union
 
 import torch
+from huggingface_hub import snapshot_download
```
If this package wasn't directly imported before in this codebase until this change, make sure it's accounted for in a requirements.txt file somewhere (even if it's already indirectly required as a dependency of another package).

sangstar replied:
`huggingface_hub` is directly imported fairly frequently in the codebase.

Eta0 commented on vllm/model_executor/model_loader/tensorizer.py:
```diff
@@ -76,6 +79,7 @@ def __post_init__(self):
         # check if the configuration is for a sharded vLLM model
         self._is_sharded = isinstance(self.tensorizer_uri, str) \
             and re.search(r'%0\dd', self.tensorizer_uri) is not None
+        self.tensorizer_dir = self.tensorizer_uri.rpartition("/")[0]
```
You can use `os.path.dirname` if that is what you intend to capture here

Eta0 commented on vllm/model_executor/model_loader/tensorizer.py:
```diff
@@ -11,6 +12,8 @@
 
 import torch
+from huggingface_hub import snapshot_download
+from safetensors.torch import load_file
```
IMO this should be imported as `import safetensors.torch` and referenced later as `safetensors.torch.load_file`, because seeing it invoked as a function simply named `load_file()` later on looks very vague, whereas seeing it invoked as `safetensors.torch.load_file()` would give more useful context.

Eta0 commented on vllm/model_executor/model_loader/tensorizer.py:
```diff
@@ -466,3 +470,43 @@ def tensorize_vllm_model(engine_args: EngineArgs,
+    """
+    lora_files = snapshot_download(repo_id=lora_path)
+
+    # Current LoRA loading logic in
+    # vllm.lora.models.LoRAModel.from_local_checkpoint assumes that
+    # the tensors and config filenames are adapter_model.safetensors and
+    # adapter_config.json respectively, so this logic makes the same
+    # assumption
+    tensor_path = os.path.join(lora_files, "adapter_model.safetensors")
+    config_path = os.path.join(lora_files, "adapter_config.json")
```
If this is something that could be subject to change, instead of commenting this here, you could refactor `vllm.lora.models.LoRAModel.from_local_checkpoint` so that it uses a helper function to retrieve these paths, and then use the same helper function here. This would make the link between the two usages more clear (e.g. an IDE could tell someone looking at it that it's used in multiple places, or if it's edited in a way that makes it incompatible, the test suite here will more likely raise a red flag).

As pseudocode:
```python
# In vllm.lora.models

def lora_file_paths(lora_dir_path):
    # Instantiate a dataclass, namedtuple, etc. with all the assumed filenames
    foo_path = os.path.join(...)
    bar_path = os.path.join(...)
    # ...Others...
    return ...

def from_local_checkpoint(lora_dir, ...):
    lora_files = lora_file_paths(lora_dir)  # Instead of a bunch of os.path.join calls
    # Then the rest of the loading logic


# Here, in vllm.model_executor.model_loader.tensorizer

from vllm.lora.models import lora_file_paths

def tensorize_lora_adapter(...):
    lora_dir = ...
    lora_files = lora_file_paths(lora_dir)
    # Then the rest of the serialization logic
```

On the other hand, if these names are set in stone and have reason they probably will not change, then don't worry about it.

sangstar replied:
`adapter_model.bin`, `adapter_model.safetensors` and `adapter_config.json` are all values that are set in stone, yep. LoRA adapters on HF Hub have that naming convention.

Eta0 commented on vllm/model_executor/model_loader/tensorizer.py:
```diff
@@ -466,3 +470,43 @@ def tensorize_vllm_model(engine_args: EngineArgs,
+
+    tensorizer_args = tensorizer_config._construct_tensorizer_args()
+
+    with open_stream(f"{tensorizer_config.tensorizer_dir}/adapter_config.json",
+                     mode="wb+",
+                     **tensorizer_args.stream_params) as f:
+
+        f.write(json.dumps(config).encode("utf-8"))
+
+    lora_uri = (f"{tensorizer_config.tensorizer_dir}"
+                f"/adapter_model.tensors")
```
As the dual to the preceding comment, these paths should also ideally be defined in a helper function so that the deserialization logic can share them without implicitly relying on using the same copy-pasted, hardcoded names.

sangstar replied:
Given that LoRA adapter model artifact filenames either start with `adapter_model.*` or `adapter_config.json`, I felt it would be fine to just hard-code and force filenames for serialized `adapter_model.safetensors` files to be `adapter_model.tensors`.

Eta0 commented on tests/entrypoints/openai/test_tensorizer_entrypoint.py:
```diff
@@ -0,0 +1,114 @@
+        "[user] Write a SQL query to answer the question based on the table schema.\n\n context: CREATE TABLE table_name_11 (nationality VARCHAR, elector VARCHAR)\n\n question: When Anchero Pantaleone was the elector what is under nationality? [/user] [assistant]",  # noqa: E501
+    ]
+
+    llm.generate(prompts,
+                 sampling_params,
+                 lora_request=LoRARequest("sql-lora",
+                                          1,
+                                          tmp_dir.name,
+                                          tensorizer_config=TensorizerConfig(
+                                              tensorizer_uri=model_uri)))
```
What does this test? It feels like this function has an elaborately-crafted prompt setup but doesn't end up using it for anything specific other than seeing if it runs without raising an exception.

Is the extent of this test case just checking that these functions can be called without runtime errors? If so, I'd simplify the prompt a bit, and use a less-intense model to test with. Let me know if there's more nuance here I'm not seeing, though. I'm not 100% on how this feature works.

sangstar replied:
The LoRA adapter is only actually loaded in to the model when the `.generate` method is called for an `LLM` with a `lora_request: LoRARequest` kwarg, and yes, it's to try to make sure no exceptions are raised. Since this is the testing script I'm asking to be included in to their fast checks (which are always ran), I thought it would be more desirable for the developers to sequester more fine-grained unit tests in the dedicated `tests/tensorizer_loader.py` tests and have these tests just ensure end-to-end use doesn't result in any exceptions.

It was sort of inspired by other tests that are included in the fast check suite, such as `test_basic_correctness.py`, which are almost entirely void of any `assert` statements and just make sure end-to-end API stuff works.

Eta0 commented on tests/entrypoints/openai/test_tensorizer_entrypoint.py:
```diff
@@ -0,0 +1,114 @@
+
+from ...utils import RemoteOpenAIServer
+
+MODEL_NAME = "meta-llama/Llama-2-7b-hf"
+LORA_PATH = "yard1/llama-2-7b-sql-lora-test"
```
`meta-llama/Llama-2-7b-hf` as distributed on Hugging Face is a gated-access model that generally requires a token with special approval. Is that still the case in the way it is used here? Is there another model that could be used for testing that is not as complicated to access?

And, additionally, is it necessary to use a 7B model here? There are many benefits to keeping test cases lightweight (e.g. making it easier to run them on CI machines with limited compute resources, getting the test suite to finish faster, and so on). If this can be replaced with a smaller model without invalidating the tests, that would likely be less of a burden for project maintenance long-term.

sangstar replied:
Like I mentioned in a previous comment, vLLM's CI runners have an access token authenticated to download Llama 2, and is loaded fairly often in their testing suite, so I figured it was reasonable to hold them to being able to handle this.

That said, the point of this test script was to make sure we add a tensorizer test script to their fast check tests, so it might be a good idea to prioritize a smaller model size and possibly try to find a different LoRA adapter, because they may not like a model size this large having to be loaded for their fast check tests, unless this is already being done, so I take your point and think I need to make sure this is absolutely necessary.

Eta0 commented on examples/other/tensorize_vllm_model.py:
```diff
@@ -169,11 +180,44 @@ def parse_args():
-    )
+    if args.lora_path:
+        from vllm import SamplingParams
+        from vllm.lora.request import LoRARequest
```
Do these need to be local, conditional imports? Unless importing a module has a substantial performance cost or requires extra dependencies that may not be met, it would be better to put them at the top of the module in the usual block of imports.

The import from `vllm` is definitely safe to make unconditional; there is even already another `from vllm import ...` statement at the top of the file, so this will definitely not have any overhead or extra requirements.

sangstar replied:
There's no need for conditional imports here. I'll put these at the top.

Eta0 commented on examples/other/tensorize_vllm_model.py:
```diff
@@ -107,6 +107,17 @@ def parse_args():
+        help="Path to a LoRA adapter to "
+        "serialize along with model tensors. This can then be deserialized "
+        "along with the model by passing a tensorizer_config kwarg to "
+        "LoRARequest with type TensorizerConfig."
```
The note about deserialization is somewhat hard to understand. It may be helpful to refer to example code or documentation here.

If nothing else, some more information on usage or a brief example could be included in the argument parser description or [epilog](https://docs.python.org/3/library/argparse.htl#epilog), though in general that is not the most accessible spot to sequester useful information.

Eta0 commented on vllm/lora/models.py:
```diff
@@ -179,20 +179,21 @@ def from_lora_tensors(
+                              embedding_padding_modules: Optional[
+                                  List[str]] = None,
+                              weights_mapper: Optional[WeightsMapper] = None,
+                              **kwargs) -> "LoRAModel":
```
Update the docstring to reflect any added parameters.

Eta0 commented on vllm/lora/request.py:
```diff
@@ -31,6 +35,7 @@ class LoRARequest(
     lora_local_path: Optional[str] = msgspec.field(default=None)
     long_lora_max_len: Optional[int] = None
     base_model_name: Optional[str] = msgspec.field(default=None)
+    tensorizer_config: Optional["TensorizerConfig"] = None
```
Document the usage of this somewhere, e.g. the docstring for this class.

Eta0 commented on .buildkite/test-pipeline.yaml:
```diff
@@ -222,7 +222,7 @@ steps:
     - python3 offline_inference/vision_language.py
     - python3 offline_inference/vision_language_multi_image.py
-    - python3 other/tensorize_vllm_model.py --model facebook/opt-125m serialize --serialized-directory /tmp/ --suffix v1 && python3 other/tensorize_vllm_model.py --model facebook/opt-125m deserialize --path-to-tensors /tmp/vllm/facebook/opt-125m/v1/model.tensors
+    - python3 other/tensorize_vllm_model.py --model meta-llama/Llama-2-7b-hf --lora-path yard1/llama-2-7b-sql-lora-test serialize --serialized-directory /tmp/ --suffix v1 && python3 other/tensorize_vllm_model.py --model meta-llama/Llama-2-7b-hf --lora-path yard1/llama-2-7b-sql-lora-test deserialize --path-to-tensors /tmp/vllm/meta-llama/Llama-2-7b-hf/v1/model.tensors
```
It's very plausible that a CI pipeline would not have the requisite permissions to access a gated `meta-llama` model. As mentioned in another comment, the choice of model here should be reconsidered for access and for size.

sangstar replied:
Good point. The choice of model here wasn't arbitrary, but rather due to the fact that any test that vLLM runs that tests its LoRA machinery invariably uses the `yard1/llama-2-7b-sql-lora-test` LoRA adapter for `meta-llama/Llama-2-7b-hf`. vLLM's CI suite has a token authenticated to download Llama 2 weights and Llama 2 is used frequently for vLLM's tests ([an example for a LoRA test explicitly](https://github.com/vllm-project/vllm/blob/main/tests/lora/test_add_lora.py)), so the runners are capable of loading it without any OOM issues. I felt there was no reason to break from the conventional choice of LoRA adapter and model here, and would give the developers greater confidence for the integration if it's working for their chosen adapter.

sangstar committed:
Apply suggestions from code review

Includes a few style fixes, a correction on not overcomplicating the device for `TensorDeserializer`, and improving the `tmp_dir` fixture for `test_tensorizer_entrypoint.py`.

Co-authored-by: Eta <24918963+Eta0@users.noreply.github.com>

sangstar reviewed (commented):
Just meant to respond to review comments; this is not intended as a review on my part. >.<

sangstar committed:
chore: Remove conditional imports, add `TODO`s from review

sangstar committed:
fix: Ensure proper cleanup when serializing, reduce code smell for `tmp_dir` fixture

sangstar committed:
fix: Remove `**kwargs`

sangstar committed:
fix: Specify docstring for `tensorizer_config`, fully remove kwarg dependence

sangstar committed:
feat: Introduce `lora_dir` attribute for `TensorizerConfig`
